fix(ci): Autolabel pull requests with the component names. Fix for the issue# 8531

ranj8work opened the pull request:
<!--
    Tag your PR title with the components that it touches along with
    the type of change
    E.g. "fix(orc8r): Fix reindexer race condition" or "feat(agw) ..."
-->

## Summary
fix(ci): Autolabel pull requests with the component names

<!-- Enumerate changes you made and why you made them -->
New github action workflows are added to automatically label the pull requests with the component names.
The changes are made as per the ask in the issue [8531](https://github.com/magma/magma/issues/8531)
8/23 - Code changes have been made as per the review comments - to have only one workflow file instead of one for each component.
8/24 - Fixed few review comments and tested.

## Test Plan
All the details of the testing performed are put under the comments for the issue 8531 at [here](https://github.com/magma/magma/issues/8531#issuecomment-898110736)

8/23 - Testing is done in my forked [test PR](https://github.com/ranj8work/magma/pull/20) . The github actions execution for the workdlow is [here](https://github.com/ranj8work/magma/actions/workflows/autolabel-pullrequests.yml)

8/24 The changes after addressing the review comments are retested and the pull request tested is [this](https://github.com/ranj8work/magma/pull/20)
The workflow action logs is [here](https://github.com/ranj8work/magma/actions/runs/1160549539)
The raw log of the action is [here](https://pipelines.actions.githubusercontent.com/7dmbOFWUgQKPHyNmXLGwlMrDWJiwYxhRkd3yiwwL45Oklmua4G/_apis/pipelines/1/runs/806/signedlogcontent/3?urlExpires=2021-08-24T01%3A22%3A18.0790780Z&urlSigningMethod=HMACV1&urlSignature=wnHD1qBR5QUs8LXOtWP8CII7U6YmtSSe%2B1RnjlwI4TI%3D)

<!--
    How did you test your change? How do you know it works?
    Add supporting screenshots, terminal pastes, etc. as necessary
-->

## Additional Information

- [ NO] This change is backwards-breaking

<!--
    If this is a backwards-breaking change, document the upgrade instructions.
    All upgrade instructions for backwards-breaking changes will be aggregated
    in the next release's changelog so this is very important to fill out.
-->

ranj8work committed:
github workflows to autolabel pull requests as per issue-8531

Signed-off-by: Ranjini Iyer <ranj8work@gmail.com>

ranj8work committed:
Merge https://github.com/magma/magma into autolabel-pullrequests

quentinDERORY commented:
The only components we want to labels are:

- ci
- feg
- agw
- orc8r
- cwf
- openwrt
- showtech
- xwf

No need to have workflow labeling agw-xxx

quentinDERORY reviewed (changes requested):
See comment about labels to use

ranj8work committed:
Consolidated the logic of auto labeling the pull requests in to one github workflow file

Signed-off-by: Ranjini Iyer <ranj8work@gmail.com>

ranj8work commented:
Hi @quentinDERORY and @tmdzk 
I have consolidated the logic in to one workflow file for all components instead of one for each component.
Testing is done in my forked [test PR](https://github.com/ranj8work/magma/pull/20) . The github actions execution for the workflow is [here](https://github.com/ranj8work/magma/actions/workflows/autolabel-pullrequests.yml)
Please review and let me know if you have any questions.

quentinDERORY commented on .github/workflows/autolabel-pullrequests.yml:
```diff
@@ -0,0 +1,122 @@
+              labels: lbls,
+            })
+
+
```
Please address all lint issues.

quentinDERORY commented on .github/workflows/autolabel-pullrequests.yml:
```diff
@@ -0,0 +1,122 @@
+  comment:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: hmarr/debug-action@v2
```
you can get rid of this

quentinDERORY commented on .github/workflows/autolabel-pullrequests.yml:
```diff
@@ -0,0 +1,122 @@
+      - 'xwf/**'
+
+jobs:
+  # This job is to automatically apply lte labels to pull requests
```
you can remote this comment, the name of the job is pretty explicit

quentinDERORY commented on .github/workflows/autolabel-pullrequests.yml:
```diff
@@ -0,0 +1,122 @@
+             } // end of switch case
+            } // end of for loop
+
+            let uniqLbls = new Set(lbls);
```
you could have created a set from the beginning.

ranj8work replied:
@quentinDERORY Eventually I still have to convert the Set in to an array as the API call takes only the array of strings for the labels.

quentinDERORY commented:
> Hi @quentinDERORY and @tmdzk
> I have consolidated the logic in to one workflow file for all components instead of one for each component.
> Testing is done in my forked [test PR](https://github.com/ranj8work/magma/pull/20) . The github actions execution for the workflow is [here](https://github.com/ranj8work/magma/actions/workflows/autolabel-pullrequests.yml)
> Please review and let me know if you have any questions.

In the logs of the workflow, I see a lot of undefined https://github.com/ranj8work/magma/runs/3396539428?check_suite_focus=true .

ranj8work committed:
autolabel-pullrequests.yml

Signed-off-by: Ranjini Iyer <ranj8work@gmail.com>

ranj8work commented:
@quentinDERORY - I addressed all of your comments. The yamlint is also successful.
The changes are retested and the pull request tested is [this](https://github.com/ranj8work/magma/pull/20)
The workflow action logs is [here](https://github.com/ranj8work/magma/actions/runs/1160549539)
The raw log of the action is [here](https://pipelines.actions.githubusercontent.com/7dmbOFWUgQKPHyNmXLGwlMrDWJiwYxhRkd3yiwwL45Oklmua4G/_apis/pipelines/1/runs/806/signedlogcontent/3?urlExpires=2021-08-24T01%3A22%3A18.0790780Z&urlSigningMethod=HMACV1&urlSignature=wnHD1qBR5QUs8LXOtWP8CII7U6YmtSSe%2B1RnjlwI4TI%3D)

ranj8work committed:
Removed the duplicate array of lbl strings and addressed few review comments

Signed-off-by: Ranjini Iyer <ranj8work@gmail.com>

quentinDERORY commented on .github/workflows/autolabel-pullrequests.yml:
```diff
@@ -0,0 +1,118 @@
+  comment:
+    runs-on: ubuntu-latest
+    steps:
+      # - uses: hmarr/debug-action@v2
```
delete commented code.

ranj8work replied:
done

ranj8work committed:
deleted comented line

Signed-off-by: Ranjini Iyer <ranj8work@gmail.com>

hcgatewood commented on .github/workflows/autolabel-pullrequests.yml:
```diff
@@ -0,0 +1,117 @@
+            });
+
+            for (const f of pulledFiles['data']) {
+             switch (true) {
```
1. It would be preferable to pull this logic into a separate yaml config file, then this function can be "dumb" and just read in and operate based on that yaml config file
2. Unless I'm mistaken, looks like this only applies 1 label per PR (?). We actually want all the relevant component labels for every component the pr changes

cc @ranj8work @quentinDERORY

quentinDERORY replied:
1. Instead of putting into a yaml file, it would be best to parse the CODEOWNER file directly. Tracking it here: https://github.com/magma/magma/issues/8765
2. it iterates over all the files and aggregates multiple labels to apply. Here is an example PR showing multiple labels being applied: https://github.com/magma/magma/pull/8715

hcgatewood replied:
1. Agreed, perfect sounds good
2. Ah yes I didn't parse the for-loop. Sounds good 🙂
